emu/src/memory/dma.rs

```diff
@@ -0,0 +1,124 @@
+        }
+    }
+}
+
+impl IoDevice for DmaRegisters {
+    type Address = usize;
+    type Value = u8;
+
+    fn read_at(&self, address: Self::Address) -> Self::Value {
+        match address {
+            0 => self.source_address.read().get_byte(0),
+            1 => self.source_address.read().get_byte(1),
+            2 => self.source_address.read().get_byte(2),
+            3 => self.source_address.read().get_byte(3),
+            4 => self.destination_address.read().get_byte(0),
+            5 => self.destination_address.read().get_byte(1),
+            6 => self.destination_address.read().get_byte(2),
+            7 => self.destination_address.read().get_byte(3),
+            8 => self.word_count.read().get_byte(0),
+            9 => self.word_count.read().get_byte(1),
+            10 => self.control.read().get_byte(0),
+            11 => self.control.read().get_byte(1),
+            _ => panic!("Not implemented read memory address: {address:x}"),
+        }
+    }
+
+    fn write_at(&mut self, address: Self::Address, value: Self::Value) {
+        match address {
+            0 => self.source_address.set_byte(0, value),
+            1 => self.source_address.set_byte(1, value),
+            2 => self.source_address.set_byte(2, value),
+            3 => self.source_address.set_byte(3, value),
+            4 => self.destination_address.set_byte(0, value),
+            5 => self.destination_address.set_byte(1, value),
+            6 => self.destination_address.set_byte(2, value),
+            7 => self.destination_address.set_byte(3, value),
+            8 => self.word_count.set_byte(0, value),
+            9 => self.word_count.set_byte(1, value),
+            10 => self.control.set_byte(0, value),
+            11 => self.control.set_byte(1, value),
+            _ => panic!("Not implemented write memory address: {address:x}"),
+        }
+    }
+}
+
+pub struct Dma {
+    banks: [DmaRegisters; 4],
+}
+
+impl Default for Dma {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Dma {
+    pub fn new() -> Self {
+        Self {
+            banks: [
+                DmaRegisters::default(),
+                DmaRegisters::default(),
+                DmaRegisters::default(),
+                DmaRegisters::default(),
+            ],
+        }
+    }
+}
+
+impl IoDevice for Dma {
+    type Address = usize;
+    type Value = u8;
+
+    fn read_at(&self, address: Self::Address) -> Self::Value {
+        match address {
+            0x040000B0..=0x040000BB => self.banks[0].read_at(address - 0x040000B0),
+            0x040000BC..=0x040000C7 => self.banks[1].read_at(address - 0x040000BC),
+            0x040000C8..=0x040000D3 => self.banks[2].read_at(address - 0x040000C8),
+            0x040000D4..=0x040000DF => self.banks[3].read_at(address - 0x040000D4),
+            0x040000E0 => {
+                log("read in ununsed 0x040000E0");
+                0
```

Comment:
why hardcoded 0?

Reply:
It shouldn't be used... Tried to put zero :)

Reply:
But in other memory regions we used hashmaps for these cases

Reply:
I can but it is only 1... I hope is actually not used

Reply:
Maybe we can put a single variable, I don't know it was just to be "consistent", not mandatory I guess